File: sync-bridge-nestjs/src/sync/sync.mapper.ts
import { BadRequestException, Injectable } from '@nestjs/common';
import { CustomerDto, EmployeeDto, OrderDto, ProductDto } from './dto/sync.dto';
import { Customer } from './entities/customer.entity';
import { Employee } from './entities/employee.entity';
import { Order } from './entities/order.entity';
import { OrderItem } from './entities/order-item.entity';
import { Product } from './entities/product.entity';

@Injectable()
export class SyncMapper {
  mapCustomer(dto: CustomerDto): Customer {
    const entity = new Customer();
    if (dto.id !== undefined) entity.id = dto.id;
    entity.email = dto.email;
    entity.firstName = dto.firstName;
    entity.lastName = dto.lastName;
    entity.defaultCurrency = dto.defaultCurrency ?? 'USD';
    return entity;
  }

  mapProduct(dto: ProductDto): Product {
    const entity = new Product();
    if (dto.id !== undefined) entity.id = dto.id;
    entity.name = dto.name;
    entity.description = dto.description;
    entity.price = dto.price;
    entity.currency = dto.currency ?? 'USD';
    entity.active = dto.active ?? true;
    entity.weightGrams = dto.weightGrams;
    return entity;
  }

  mapOrder(dto: OrderDto): Order {
    const entity = new Order();
    if (dto.id !== undefined) entity.id = dto.id;
    entity.orderNumber = dto.orderNumber;

    const customer = new Customer();
    customer.id = dto.customerId;
    entity.customer = customer;

    entity.status = dto.status;
    entity.currency = dto.currency ?? 'USD';

    const items: OrderItem[] = [];
    if (dto.items && dto.items.length) {
      const invalid = dto.items.some(
        (item) => item.qty === undefined || item.unitPrice === undefined,
      );
      if (invalid) {
        throw new BadRequestException(
          'Order items must include non-null qty and unit_price',
        );
      }
      const calcAmount = dto.items.reduce(
        (sum, item) => sum + item.qty * item.unitPrice,
        0,
      );

      if (dto.amount === undefined) {
        entity.amount = calcAmount;
      } else if (dto.amount !== calcAmount) {
        throw new BadRequestException(
          `Order amount must equal the sum of item prices (qty * unit_price). Calculated=${calcAmount} provided=${dto.amount}`,
        );
      } else {
        entity.amount = dto.amount;
      }

      dto.items.forEach((itemDto) => {
        const orderItem = new OrderItem();
        if (itemDto.id !== undefined) orderItem.id = itemDto.id;
        const product = new Product();
        product.id = itemDto.productId;
        orderItem.product = product;
        orderItem.qty = itemDto.qty;
        orderItem.unitPrice = itemDto.unitPrice;
        orderItem.order = entity;
        items.push(orderItem);
      });
    } else {
      if (dto.amount === undefined) {
        throw new BadRequestException('Order must include items or an amount');
      }
      entity.amount = dto.amount;
    }

    entity.items = items;
    return entity;
  }

  mapEmployee(dto: EmployeeDto): Employee {
    const entity = new Employee();
    if (dto.id !== undefined) {
      const parsed = Number(dto.id);
      if (Number.isNaN(parsed)) {
        throw new BadRequestException('Employee id must be numeric when provided');
      }
      entity.id = parsed;
    }
    entity.employeeId = dto.employeeId;
    entity.firstName = dto.firstName;
    entity.middleName = dto.middleName;
    entity.lastName = dto.lastName;
    entity.gender = dto.gender;
    entity.email = dto.email;
    entity.phoneNumber = dto.phoneNumber;
    entity.dateOfBirth = dto.dateOfBirth;
    entity.nationality = dto.nationality;
    entity.jobLevel = dto.jobLevel;
    entity.department = dto.department;
    entity.location = dto.location;
    entity.bankAccountNumber = dto.bankAccountNumber;
    entity.company = dto.company;
    entity.jobTitle = dto.jobTitle;
    entity.costCenter = dto.costCenter;
    entity.startDate = dto.startDate;
    entity.employeeStatus = dto.employeeStatus;
    entity.managerId = dto.managerId;
    entity.managerEmail = dto.managerEmail;
    entity.lastModifiedOn = dto.lastModifiedOn;
    entity.lastModified = dto.lastModified;
    return entity;
  }
}

File: sync-bridge-csharp/SyncBridgeCsharp/Services/SyncMapper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using SyncBridgeCsharp.Dtos;
using SyncBridgeCsharp.Exceptions;
using SyncBridgeCsharp.Models;

namespace SyncBridgeCsharp.Services;

public interface ISyncMapper
{
    Customer MapCustomer(CustomerDto d);
    Product MapProduct(ProductDto d);
    Order MapOrder(OrderDto d);
    Employee MapEmployee(EmployeeDto d);
}

public class SyncMapper : ISyncMapper
{
    public Customer MapCustomer(CustomerDto d)
    {
        var c = new Customer
        {
            Email = d.Email,
            FirstName = d.FirstName,
            LastName = d.LastName,
            DefaultCurrency = string.IsNullOrWhiteSpace(d.DefaultCurrency) ? "USD" : d.DefaultCurrency,
            UpdatedAt = DateTime.UtcNow
        };
        if (d.Id.HasValue)
        {
            c.Id = d.Id.Value;
        }
        return c;
    }

    public Product MapProduct(ProductDto d)
    {
        var p = new Product
        {
            Name = d.Name,
            Description = d.Description,
            Price = d.Price ?? 0,
            Currency = string.IsNullOrWhiteSpace(d.Currency) ? "USD" : d.Currency,
            Active = d.Active ?? true,
            WeightGrams = d.WeightGrams,
            UpdatedAt = DateTime.UtcNow
        };
        if (d.Id.HasValue)
        {
            p.Id = d.Id.Value;
        }
        return p;
    }

    public Order MapOrder(OrderDto d)
    {
        var o = new Order
        {
            OrderNumber = d.OrderNumber,
            CustomerId = d.CustomerId ?? 0,
            Status = d.Status,
            Currency = string.IsNullOrWhiteSpace(d.Currency) ? "USD" : d.Currency,
            UpdatedAt = DateTime.UtcNow
        };

        if (d.Id.HasValue)
        {
            o.Id = d.Id.Value;
        }

        if (d.Items != null && d.Items.Count > 0)
        {
            // Validate items have non-null qty and unit_price
            if (d.Items.Any(it => !it.Qty.HasValue || !it.UnitPrice.HasValue || !it.ProductId.HasValue))
            {
                throw new ApiException(400, "Order items must include non-null qty and unit_price");
            }

            int calculatedAmount = d.Items.Sum(it => it.Qty!.Value * it.UnitPrice!.Value);

            if (!d.Amount.HasValue)
            {
                o.Amount = calculatedAmount;
            }
            else if (d.Amount.Value != calculatedAmount)
            {
                throw new ApiException(400, $"Order amount must equal the sum of item prices (qty * unit_price). Calculated={calculatedAmount} provided={d.Amount.Value}");
            }
            else
            {
                o.Amount = d.Amount.Value;
            }

            o.Items = d.Items.Select(it => new OrderItem
            {
                Id = it.Id ?? 0,
                ProductId = it.ProductId!.Value,
                Qty = it.Qty!.Value,
                UnitPrice = it.UnitPrice!.Value,
                Order = o
            }).ToList();
        }
        else
        {
            if (!d.Amount.HasValue)
            {
                throw new ApiException(400, "Order must include items or an amount");
            }
            o.Amount = d.Amount.Value;
        }

        return o;
    }

    public Employee MapEmployee(EmployeeDto d)
    {
        var e = new Employee
        {
            EmployeeId = d.EmployeeId,
            FirstName = d.FirstName,
            MiddleName = d.MiddleName,
            LastName = d.LastName,
            Gender = d.Gender,
            Email = d.Email ?? string.Empty,
            PhoneNumber = d.PhoneNumber,
            DateOfBirth = d.DateOfBirth,
            Nationality = d.Nationality,
            JobLevel = d.JobLevel,
            Department = d.Department,
            Location = d.Location,
            BankAccountNumber = d.BankAccountNumber,
            Company = d.Company,
            JobTitle = d.JobTitle,
            CostCenter = d.CostCenter,
            StartDate = d.StartDate,
            EmployeeStatus = d.EmployeeStatus,
            ManagerId = d.ManagerId,
            ManagerEmail = d.ManagerEmail,
            LastModifiedOn = d.LastModifiedOn,
            LastModified = d.LastModified
        };

        if (long.TryParse(d.Id, out long parsedId))
        {
            e.Id = parsedId;
        }

        return e;
    }
}
